apis/objectstorage/v1alpha2/types.go

```diff
@@ -0,0 +1,336 @@
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:storageversion
+type BucketClass struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// DriverName is the name of driver associated with this bucket
+	DriverName string `json:"driverName"`
+
+	// DeletionPolicy is used to specify how COSI should handle deletion of this
+	// bucket. There are 2 possible values:
+	//  - Retain: Indicates that the bucket should not be deleted from the OSP
+	//  - Delete: Indicates that the bucket should be deleted from the OSP
+	//        once all the workloads accessing this bucket are done
+	// +kubebuilder:default:=Retain
+	DeletionPolicy DeletionPolicy `json:"deletionPolicy"`
+
+	// Quotas defines optional quotas applied to all buckets created from this Bucket Class.
+	// +optional
+	Quotas BucketQuotas `json:"quotas,omitempty"`
+
+	// Parameters is an opaque map for passing in configuration to a driver
+	// for creating the bucket
+	// +optional
+	Parameters map[string]string `json:"parameters,omitempty"`
+}
+
+type BucketQuotas struct {
+	// MaxObjects specifies the maximum number of objects that will be allowed in a bucket.
+	// +optional
+	// +nullable
+	MaxObjects *uint64 `json:"maxObjects"`
+
+	// MaxSize specifies the maximum allowed size counting all objects in the bucket. This parameter
+	// should not be assumed to be a hard limit. Most object storage providers are only able to
+	// check the total size of a bucket periodically, and high-bandwidth bucket writes may allow the
+	// total size to surpass the limit by some margin.
+	// +optional
+	// +nullable
+	MaxSize *resource.Quantity `json:"maxSize"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type BucketClassList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BucketClass `json:"items"`
+}
+
+// +genclient
+// +genclient:nonNamespaced
+// +genclient:noStatus
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:storageversion
+type BucketAccessClass struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// DriverName is the name of driver associated with
+	// this BucketAccess
+	DriverName string `json:"driverName"`
+
+	// AuthenticationType denotes the style of authentication
+	// It can be one of
+	// Key - access, secret tokens based authentication
+	// IAM - implicit authentication of pods to the OSP based on service account mappings
+	AuthenticationType AuthenticationType `json:"authenticationType"`
+
+	// Parameters is an opaque map for passing in configuration to a driver
+	// for granting access to a bucket
+	// +optional
+	Parameters map[string]string `json:"parameters,omitempty"`
```

Comment:
There should be some sort of quota offering in BucketAccessClass as well to restrict users, say if the same bucket is shared among different users which has a limit of 10Gb. If one user consumes 9Gb and otherone will have only 1Gb access remaining

Reply:
This design is specifically only for bucket quotas. Rook users have given feedback that they want bucket quotas, but they seem to not have strong desires for access/user-level quotas. Given that, and in the interest of making small, incremental changes, I have intentionally omitted BucketAccess quotas from the proposal.